### PR TITLE
refactor endpoints code

### PR DESCRIPTION
**Pull request by ramaraochavali**

### CODE REVIEW

**Comment by ramaraochavali** [comment]
@howardjohn I was playing around with the new endpoint code to see if we can reuse the common pieces between endpoint and endpoint slice. Please see if this makes sense, if it I can move forward and do similar thing for couple of other pieces. Otherwise, if you want them to be separate for some reason, I will close this PR

**Comment by hzxuzhonghu** [comment]
What's the biggest change?

**Comment by ramaraochavali** [comment]
Two main changes
1. GetProxyServiceInstances - Is moved to endpointdiscovery and handles both types, avoiding code repeation
2. Event handling logic - handle headless service and trigger eds for others is moved to endpointdiscovery - again reusing code for both types.

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 1]
```diff
@@ -15,8 +15,17 @@
...
+	"istio.io/istio/pilot/pkg/features"
 	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
 	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/schemas"
+	"istio.io/pkg/log"
+
+	v1 "k8s.io/api/core/v1"
+	discoveryv1alpha1 "k8s.io/api/discovery/v1alpha1"
+
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: nit : blank

**Comment by ramaraochavali** [comment]
@howardjohn Do you think this refactoring is useful or should we just leave it as is? I am OK either way and happy to close this if you think it is better to leave it as is.

**Comment by howardjohn** [comment]
sorry haven't had a chance to look, not working full time the next few days

In the near future we need to read from both endpoints and endpoint slice
so we should make sure we support that

**Comment by ramaraochavali** [comment]
>> sorry haven't had a chance to look, not working full time the next few days

No problem. I was expecting that because of holidays :-). PTAL whenever you get time. Obviously not urgent. 

>> In the near future we need to read from both endpoints and endpoint slice
so we should make sure we support that

Yes. It is supported. It just refactors in such a way we have code common across the two.

**Comment by ramaraochavali** [comment]
@howardjohn PTAL, when you get time.

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 2]
```diff
@@ -28,4 +36,76 @@ type kubeEndpointsController interface {
...
+			proxyServiceInstances = getProxyServiceInstancesByEndpoint(c, *item, proxy)
+		case *discoveryv1alpha1.EndpointSlice:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpointSlice(c, item, proxy)
+		}
+		if namespace == proxyNamespace {
+			sameNamespaceEndpoints = append(sameNamespaceEndpoints, proxyServiceInstances...)
+		} else {
+			if otherNamespaceEndpoints == nil {
+				otherNamespaceEndpoints = make([]*model.ServiceInstance, 0)
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: not needed, you can do `append(nil, ...)`. L50 is technically not needed either.

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 3]
```diff
@@ -28,4 +36,76 @@ type kubeEndpointsController interface {
...
+func (e *kubeEndpoints) Run(stopCh <-chan struct{}) {
+	e.informer.Run(stopCh)
+}
+
+func (e *kubeEndpoints) handleEvent(epc kubeEndpointsController, name string, namespace string, event model.Event, obj interface{}) error {
+	log.Debugf("Handle event %s for endpoint %s in namespace %s", event, name, namespace)
+
+	// headless service cluster discovery type is ORIGINAL_DST, we do not need update EDS.
+	if features.EnableHeadlessService.Get() {
+		if obj, _, _ := e.c.services.GetIndexer().GetByKey(kube.KeyFunc(name, namespace)); obj != nil {
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: As far as I can tell this is not right. `name` here is the name of the endpoint object, right? It needs to be the name of the Service

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 4]
```diff
@@ -28,4 +36,76 @@ type kubeEndpointsController interface {
...
+			svc := obj.(*v1.Service)
+			// if the service is headless service, trigger a full push.
+			if svc.Spec.ClusterIP == v1.ClusterIPNone {
+				e.c.xdsUpdater.ConfigUpdate(&model.PushRequest{
+					Full:              true,
+					NamespacesUpdated: map[string]struct{}{namespace: {}},
+					// TODO: extend and set service instance type, so no need to re-init push context
```
Comment (howardjohn) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: not sure we should make optimizations like this during refactoring

**Comment by howardjohn** [comment]
fyi, this will conflict with https://github.com/istio/istio/pull/19839

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 5]
```diff
@@ -15,8 +15,16 @@
 package controller
 
 import (
+	"istio.io/istio/pilot/pkg/features"
 	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pilot/pkg/serviceregistry/kube"
 	"istio.io/istio/pkg/config/labels"
+	"istio.io/istio/pkg/config/schemas"
+	"istio.io/pkg/log"
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: put istio at last

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go (hunk 3 above)
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: good catch, for endpointslice it is not right

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go (hunk 4 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: Sorry. Did not get this. I did not make any optimization here. Just moved the same code from both endpoint and enpoint slice here. Can you elaborate more on this?

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go (hunk 3 above)
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: Yeah. thank. Will fix it.

**Comment by ramaraochavali** [comment]
>  At first glance, there is a lot of cyclic import within the three golang files. I do not feel very comfortable with this model. Can you separate them?

Can you elaborate more on this? `endpoints.go` and `endpointslice.go` depend on `endpointsdiscovery.go`. Not the other way.  Please share if you have any suggestions on separating them further. Happy to implement.,

File: pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go [hunk 6]
```diff
@@ -28,4 +36,73 @@ type kubeEndpointsController interface {
...
+	for _, item := range e.informer.GetStore().List() {
+		var proxyServiceInstances []*model.ServiceInstance
+		namespace := ""
+		switch item := item.(type) {
+		case *v1.Endpoints:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpoint(c, *item, proxy)
+		case *discoveryv1alpha1.EndpointSlice:
+			namespace = item.Namespace
+			proxyServiceInstances = getProxyServiceInstancesByEndpointSlice(c, item, proxy)
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: I mean these code, they still need to know the specific type and call the specific implementation.

Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/kube/controller/endpointsdiscovery.go]: I see. Made some changes to separate them out. See if the new commit makes sense.

**Comment by ramaraochavali** [comment]
@hzxuzhonghu refactored a little bit. Please see if this makes sense

**Comment by ramaraochavali** [comment]
@howardjohn Do you have any further comments? There is an open comment here https://github.com/istio/istio/pull/19713#discussion_r362539790?

**Comment by ramaraochavali** [comment]
no problem. thank you

**Pull request merged**